wagtail_ab_testing/models.py

```diff
@@ -8,20 +12,157 @@ def get_current_for_page(self, page):
 
 
 class AbTest(models.Model):
+    """
+    Represents an A/B test that has been set up by the user.
+
+    The live page content is used as the control, the revision pointed to in
+    the `.treatment_revision` field contains the changes that are being tested.
+    """
+
     class Status(models.TextChoices):
         DRAFT = 'draft', __('Draft')
         RUNNING = 'running', __('Running')
         PAUSED = 'paused', __('Paused')
         CANCELLED = 'cancelled', __('Cancelled')
         COMPLETED = 'completed', __('Completed')
 
+    class Variant(models.TextChoices):
+        CONTROL = 'control', __('Control')
+        TREATMENT = 'treatment', __('Treatment')
+
     page = models.ForeignKey('wagtailcore.Page', on_delete=models.CASCADE, related_name='ab_tests')
     name = models.CharField(max_length=255)
-    variant_revision = models.ForeignKey('wagtailcore.PageRevision', on_delete=models.CASCADE, related_name='+')
+    treatment_revision = models.ForeignKey('wagtailcore.PageRevision', on_delete=models.CASCADE, related_name='+')
     goal_type = models.CharField(max_length=255)
     # TODO Page chooser
     goal_page = models.ForeignKey('wagtailcore.Page', null=True, blank=True, on_delete=models.SET_NULL, related_name='+')
     sample_size = models.PositiveIntegerField()
     status = models.CharField(max_length=20, choices=Status.choices, default=Status.DRAFT)
 
     objects = AbTestManager()
+
+    def finish(self, *, cancel=False):
+        """
+        Finishes the test.
+        """
+        self.status = self.Status.CANCELLED if cancel else self.Status.COMPLETED
+
+        self.save(update_fields=['status'])
+
+    def add_participant(self, variant=None):
+        """
+        Inserts a new participant into the log. Returns the variant that they should be shown.
+        """
+        # Get current numbers of participants for each variant
+        stats = self.hourly_logs.aggregate(
+            control_participants=Sum('participants', filter=Q(variant=self.Variant.CONTROL)),
```

Comment:
If we're doing filtering here, we're adding a WHERE clause the the SQL statement I believe. If this is the case, should we consider adding `db_index` to the relevant fields? Or would this likely never get big enough to warrant indexing the columns?

Reply:
Good point, thanks! I'll add this in a subsequent PR!

Reply:
The existing ``unique_together`` on ``('ab_test', 'variant', 'date', 'hour'),`` should cover this. It's filtering by ``ab_test`` and ``variant`` which is a prefix of this index so it should get used for this query.